include/lsst/afw/geom/Endpoint.h

```diff
@@ -0,0 +1,303 @@
+        if (nAxes != this->getNAxes()) {
+            std::ostringstream os;
+            os << "number of axes provided " << nAxes << " != " << this->getNAxes() << " required";
+            throw std::invalid_argument(os.str());
+        }
+    }
+
+    int _getNAxes(ndarray::Array<double, 2, 2> const & data) const {
+        return data.getSize<1>();
+    }
+
+    int _getNAxes(ndarray::Array<double, 1, 1> const & data) const {
+        return data.getSize<0>();
+    }
+
+    int _getNAxes(std::vector<double> const & data) const {
+        return data.size();
+    }
+
+    int _getNPoints(ndarray::Array<double, 2, 2> const & data) const {
+        return data.getSize<0>();
+    }
+
+private:
+    int _nAxes;  /// number of axes
+};
+
+/**
+Base class for endpoints with Array = std::vector<Point>
+*/
+template <typename PointT>
+class BaseVectorEndpoint : public BaseEndpoint<PointT, std::vector<PointT>> {
+public:
+    using Array = std::vector<PointT>;
+    using Point = PointT;
+
+    explicit BaseVectorEndpoint(int nAxes) : BaseEndpoint<Point, Array>(nAxes) {};
+
+    virtual ~BaseVectorEndpoint() {};
+
+    virtual int getNPoints(Array const & arr) const override {
+        return arr.size();
+    }
+
+    virtual std::vector<double> dataFromPoint(Point const & point) const override;
+
+    virtual ndarray::Array<double, 2, 2> dataFromArray(Array const & arr) const override;
+
+    virtual Point pointFromData(std::vector<double> const & data) const override;
+
+    virtual Array arrayFromData(ndarray::Array<double, 2, 2> const & data) const override;
+};
+
+/**
+A generic endpoint for data in the format used by ast::Mapping
+
+Thus supports all ast frame classes and any number of axes, and thus can be used as an endpoint
+for any ast::Mapping.
+*/
+class GenericEndpoint: public BaseEndpoint<std::vector<double>, ndarray::Array<double, 2, 2>> {
+public:
+    /// Construct a GenericEndpoint with the specified number of axes
+    explicit GenericEndpoint(int nAxes) : BaseEndpoint(nAxes) {};
+
+    virtual ~GenericEndpoint() {};
+
+    virtual int getNPoints(Array const & arr) const override {
+        return arr.getSize<0>();
+    }
+
+    virtual std::vector<double> dataFromPoint(Point const & point) const override;
+
+    virtual ndarray::Array<double, 2, 2> dataFromArray(Array const & arr) const override;
+
+    virtual Point pointFromData(std::vector<double> const & data) const override;
+
+    virtual Array arrayFromData(ndarray::Array<double, 2, 2> const & data) const override;
+};
+
+
+/**
+An endpoint for Point<double, N>
+
+@tparam N  number of axes; the only allowed values are 2 or 3
+*/
+template <int N>
+class PointEndpoint : public BaseVectorEndpoint<Point<double, N>> {
+public:
+    /// Construct a PointEndpoint; template parameter N specifies the number of axes
+    explicit PointEndpoint() : BaseVectorEndpoint<Point<double, N>>(N) {}
+
+    /// This constructor is used by Transform; nAxes must be 2
```

Comment:
The note is wrong (a holdover from when I was only supporting Point2D); thanks for catching this.